### ISSUE TITLE
How to compile / use

### ISSUE THREAD

**Issue by NNNenov** [sylefeb/VoxModSynth#1]
Hello,

I am extremely interested in trying this out, as the rule creation looks extremely intuitive and easy!

I have some questions about how to use the project:

 1 - I don't know any C++, does this project need to be complied in order to be used, or is it meant to be used via C++ environment, like Visual Studio?

I feel like I understood the rule definition very well, but I did not understand how to actually use it:

 2 - Are there configuration file(s) where you define tile size (tile dimensions and final output dimensions)?
		
 3 - In the rule file, where adjacencies are defined, the colours do not cover orientation, 
		so am I right in thinking that you need a different colour for each orientation of the same tile?
		
 4 - How are the links to the high resolution tiles defined? I noticed theres a "castle_detailed.slab.vox" file.
		Does this mean, you need to make a file with the same name as the adjecency rule file, but append "_detailed" and place it in the same "exemplars" folder?
			
 5 - inside the _detailed vox file is an unusual distribution of high res tiles, could you elaborate on their placement?
		Or can they be placed *anywhere* (or specific increments) as long as the colour matches the rule colour index and (do they) need to all conform to one tile resoluton?
		
 6 - Am I right to assume this program therefore does not support "textured" voxel tiles? as the colour is reserved for ID?

**Comment by sylefeb** [comment]
Hi NNNenov,

1) I compiled it both under Windows (Visual Studio 2013) and under Linux with gcc. The project is using CMake and standard C++ without dependencies, so it should be possible to compile it on many platforms.

You need to use CMake to generate the project files for Visual Studio: https://cmake.org/
I found this video that seems to cover the basics (but there are likely tons of other tutorials): https://www.youtube.com/watch?v=LxHV-KNEG3k

After producing the project files, simply open with Visual Studio and press "F5" to compile and run.

2) No: this is all deduced from the sizes of the input voxel data. There are some configuration variables at the top of the code, these are described with comments.

3) Yes, if necessary (sometimes it is ok to reuse the same colors, sometimes you want to distinguish). This is a downside of this approach, for cases with many symmetries it can become a bit tedious. However it enables a finer control. Note that from the code point of view, tiles with different orientations are indeed different: internally they do correspond to different labels.

4) Yes, you have to make another file, which size is an exact multiple of the 'low res' one. For instance, 'castle.slab' is 8x8x12 and 'castle_detailed.slab' is 64x64x96 because the tiles are 8x8x8 voxels.
Note that 'castle.slab.vox' and 'castle_detailed.slab.vox' correspond to each other: there is a single voxel in 'castle.slab.vox' for each detailed tile in 'castle_detailed.slab.vox'. These are not the 'rule file' but a tilemap. The rule file is a third, different file, for instance: 'towers.slab.vox'. The voxel colors in  'castle.slab.vox' (low res) have to match the colors used in the rule file => this way we can locate the tiles in the tilemap. Colors in the detailed tiles do not matter (the code currently overrides them, but they could also be carried over to add some details to the output).
So in summary, a typical case has three files:
a- the adjacency rules (towers.slab.vox)
b- the low res tile specification (castle.slab.vox)
c- the high res tile specification (castle_detailed.slab.vox)
Colors in a and b correspond, positions in b and c correspond. This is how the code can build the detailed output from a synthesized result.

5) Hopefully answered above - let me know if that isn't clear (there are also additional comments in the code).

6) Yes for now (see line 632, "uchar solid_color = 246;". this is the used color in the current implementation). However, nothing would prevent copying colors from detailed voxels -- these can be freely chosen (I did not do that as I was focusing on 3D printing).

I hope you'll be able to use the code in your projects!
Best,
Sylvain

**Comment by NNNenov** [comment]
Hi Sylvain,

Thank you, I didn't realize there was 3 files, this makes much more sense now!

Thanks for the link, I was hoping I could compile without VS, but it sounds straightforward enough.

**Comment by NNNenov** [comment]
I tried the tutorial you linked, but when I got to the stage of "configuring" CMake, it threw this error:
![image](https://user-images.githubusercontent.com/9900819/29006808-fd3d2624-7aee-11e7-8e1e-af3e91cceafd.png)


I will try to investigate, I am a total n00b with compilers / environments / git / anything C

**Comment by sylefeb** [comment]
Strange. I just upgraded to CMake 3.9, did everything from scratch, and it worked.

Basically after specifying the folders, simply press "Configure", select "Visual Studio 12 2013 Win64". CMake will test the compiler, and then you can click "Generate". After that the project files are available (you can even open visual studio by clicking "Open Project" in CMake GUI).

Before any of that, maybe make sure you can launch Visual Studio? From your screenshot it may be an issue with the compiler. My log says:

The C compiler identification is MSVC 18.0.31101.0
The CXX compiler identification is MSVC 18.0.31101.0
Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/x86_amd64/cl.exe
Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/x86_amd64/cl.exe -- works
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/x86_amd64/cl.exe
Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 12.0/VC/bin/x86_amd64/cl.exe -- works
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Detecting CXX compile features
Detecting CXX compile features - done
Configuring done
Generating done

**Comment by NNNenov** [comment]
oh man, yeah I didn't have VS installed.. how embarrassing!

**Comment by NNNenov** [comment]
Ok, Installed Visual Studio 2013 Community and successfully made the project using CMake!
I opened  the project in VS and press F5, but got a file missing error:
![image](https://user-images.githubusercontent.com/9900819/29008607-b7b007f2-7b11-11e7-988f-e914c093df10.png)

**Comment by sylefeb** [comment]
Almost there! This is because the "ALL_BUILD" project is selected as active (see right toolbar, called "solution explorer"). In the solution explorer, select "VoxModSynth", right click, in the popup menu select "Set as startup project". Then hit F5 again, the program will run. Before you do though, switch the compiler to release mode: in the top bar, just below the TEST/ANALYSE menus, you will see a "Debug" item. Click on this, a menu will unroll, select "Release". Then instead of "F5", hit "Ctrl+F5". This will compile in "Release" mode, with full optimization, and will launch without the debugger at full speed.

**Comment by NNNenov** [comment]
Yaay! Thanks for bearing with me.  Its working! 
![image](https://user-images.githubusercontent.com/9900819/29049459-f8467a82-7bcd-11e7-8e05-869fa09318ec.png)

I was a little confused at first where the files were being read from / generated, then I realised the file paths are not relative to the .exe but rather absolute set when compiling it?

Hopefully I will have some time to actually make some tiles and rules soon!

**Comment by sylefeb** [comment]
Excellent news!! 
Yes, the paths are generated when you run the CMake script. This is the simplest way I found to deal with paths in self-contained, small code projects.
Looking forward to see what you can produce with it!

**Comment by NNNenov** [comment]
cool, thanks again for your help!
I am just playing around with rules for now to see what kind of forms I can muster.
![image](https://user-images.githubusercontent.com/9900819/29100095-5be840f6-7ca1-11e7-898b-d95255c79f09.png)

I am wondering, is there a way to control density, like, how much empty space to preserve? it seems like it is set to 1 block?

**Comment by aman-tiwari** [comment]
@NNNenov did you figure out how to modify density? All of the forms I generate seem to be far denser than the examples

**Comment by NNNenov** [comment]
I didn't play around with it much after this point, I might have a look at it again soon. I believe, although I haven't experimented much, "density" all comes down to the way you structure your example scene, it takes a lot of unique and well placed colors to make sure you don't

**Comment by NNNenov** [comment]
.. flood the scene. ( On my phone sorry)
